linux_container.go

```diff
@@ -83,36 +96,217 @@ func (c *linuxContainer) Stats() (*ContainerStats, error) {
+	child.Close()
+	if err != nil {
+		return newGenericError(err, SystemError)
+	}
+	wait := func() (*os.ProcessState, error) {
+		ps, err := cmd.Process.Wait()
+		// we should kill all processes in cgroup when init is died if we use
+		// host PID namespace
+		if !c.config.Namespaces.Contains(configs.NEWPID) {
+			c.killAllPids()
+		}
+		return ps, newGenericError(err, SystemError)
+	}
+	terminate := func(terr error) error {
+		// TODO: log the errors for kill and wait
+		cmd.Process.Kill()
+		wait()
+		return terr
+	}
+	started, err := system.GetProcessStartTime(cmd.Process.Pid)
+	if err != nil {
+		return terminate(err)
+	}
+	// Do this before syncing with child so that no children
+	// can escape the cgroup
+	if err := c.cgroupManager.Apply(cmd.Process.Pid); err != nil {
+		return terminate(err)
+	}
+	defer func() {
+		if err != nil {
+			c.cgroupManager.Destroy()
+		}
+	}()
+	var networkState configs.NetworkState
+	if err := c.initializeNetworking(cmd.Process.Pid, &networkState); err != nil {
+		return terminate(err)
+	}
+	process := processArgs{
+		Args:         args,
+		Config:       c.config,
+		NetworkState: &networkState,
+	}
+	// Start the setup process to setup the init process
+	if c.config.Namespaces.Contains(configs.NEWUSER) {
+		if err = executeSetupCmd(cmd.Args, cmd.Process.Pid, c.config, &process, &networkState); err != nil {
+			return terminate(err)
+		}
+	}
+	// send the state to the container's init process then shutdown writes for the parent
+	if err := json.NewEncoder(parent).Encode(process); err != nil {
+		return terminate(err)
+	}
+	// shutdown writes for the parent side of the pipe
+	if err := syscall.Shutdown(int(parent.Fd()), syscall.SHUT_WR); err != nil {
+		return terminate(err)
+	}
+	// wait for the child process to fully complete and receive an error message
+	// if one was encoutered
+	var ierr *initError
+	if err := json.NewDecoder(parent).Decode(&ierr); err != nil && err != io.EOF {
+		return terminate(err)
+	}
+	if ierr != nil {
+		return terminate(ierr)
+	}
+
+	c.state.InitPid = cmd.Process.Pid
+	c.state.InitStartTime = started
+	c.state.NetworkState = networkState
+	c.state.CgroupPaths = c.cgroupManager.GetPaths()
+
+	return nil
+}
+
+func (c *linuxContainer) Destroy() error {
+	status, err := c.Status()
+	if err != nil {
+		return err
+	}
+	if status != configs.Destroyed {
+		return newGenericError(nil, ContainerNotStopped)
+	}
+	return os.RemoveAll(c.root)
+}
+
+func (c *linuxContainer) Pause() error {
+	return c.cgroupManager.Freeze(configs.Frozen)
+}
+
+func (c *linuxContainer) Resume() error {
+	return c.cgroupManager.Freeze(configs.Thawed)
+}
+
+func (c *linuxContainer) Signal(signal os.Signal) error {
+	glog.Infof("sending signal %d to pid %d", signal, c.state.InitPid)
+	panic("not implemented")
+}
+
+func (c *linuxContainer) Wait() (syscall.WaitStatus, error) {
+	var status syscall.WaitStatus
+	// TODO : close exec.Cmd pipes, fix in master
+	_, err := syscall.Wait4(c.state.InitPid, &status, 0, nil)
```

Comment:
I don't like Wait4 here for a few reasons.
* it can be used only once, then it will return an error.
* we provide pid to an user, so he can call wait4 too.
* it doesn't work for containers without pidns.
We can try to use a cgroup release agent here...
I don't suggest to fix this issue right now, It's just for information. Maybe we need to add todo here.